src/acl_mem.cpp

```diff
@@ -405,6 +406,339 @@ int acl_bind_buffer_to_device(cl_device_id device, cl_mem mem) {
   return 1;
 }
 
+/**
+ * Read <size> bytes of data from device global
+ *
+ * This function extract the device global address with given name, create
+ * kernel from the given program, create temporary device usm pointer to hold
+ * data that will be copied from device global. Launch the copy kernel with
+ * correct pointers set as src and destination, then enqueue a memory copy
+ * operation that depend on copy kernel to copy from temporary device usm
+ * pointer into user provided host pointer. Then register a callback to release
+ * the necessary events, kernels, memories.
+ *
+ * @param command_queue the queue system this copy kernel will belong
+ * @param program contains copy kernel
+ * @param name name of device global, used to look up for device global address
+ * in autodiscovery string
+ * @param blocking_read whether the operation is blocking or not
+ * @param size number of bytes to read / write
+ * @param offset offset from the extracted address of device global
+ * @param ptr pointer that will hold the data copied from device global
+ * @param num_events_in_wait_list number of event that copy kernel depend on
+ * @param event_wait_list events that copy kernel depend on
+ * @param event the info about the execution of copy kernel will be stored in
+ * the event
+ * @return status code, CL_SUCCESS if all operations are successful.
+ */
+ACL_EXPORT
+CL_API_ENTRY cl_int clEnqueueReadGlobalVariableINTEL(
+    cl_command_queue command_queue, cl_program program, const char *name,
+    cl_bool blocking_read, size_t size, size_t offset, void *ptr,
+    cl_uint num_events_in_wait_list, const cl_event *event_wait_list,
+    cl_event *event) {
+  cl_int status;
+
+  // Get copy kernel from the program
+  cl_kernel kernel =
+      clCreateKernelIntelFPGA(program, "copy_device_global", &status);
+  if (status != CL_SUCCESS) {
+    return status;
+  }
+
```

Comment:
You can insert an `assert(kernel)` after checking the status to appease Klocwork.